Adding Nitrogen ThermoLibrary, Benson Groups, and HBIs

davidfarinajr opened the pull request:
- Added a new ThermoLibrary (Nitrogen_G4) of systematically generated small nitrogen-containing species. Species were calculated using Gaussian G4 method with bond additivity corrections derived from known ATcT species.
- Derived new nitrogen Group Additivity Values and HBI groups from the Nitrogen_G4 library.  Groups that already had data were held constant.

davidfarinajr commented:
Group Additivity Estimate of Nitrogen_G4 library species before new groups (with current master):
[Nitrogen_G4_GAV_Estimate_Parity_Original.pdf](https://github.com/ReactionMechanismGenerator/RMG-database/files/4010913/Nitrogen_G4_GAV_Estimate_Parity_Original.pdf)

Group Additivity Estimate of Nitrogen_G4 library species after new groups (with nitrogen branch):
[Nitrogen_G4_GAV_Estimate_Parity.pdf](https://github.com/ReactionMechanismGenerator/RMG-database/files/4010914/Nitrogen_G4_GAV_Estimate_Parity.pdf)

alongd commented:
Thanks! Could you also PR the BAC values for G4 to Arkane?

alongd commented on input/thermo/groups/group.py:
```diff
@@ -45217,19 +45217,20 @@
 entry(
     index = 1816,
-    label = "N3s-N3",
+    label = "N3s-N",
```
The `N3s-NCC` group is one of RMG's magics: It has zeroes for data, and its functionality is unclear to me. I would either change the `N3s-NCC` group to be `N3s-N3CC` to fit this specific branch in the tree, or remove it (although we should first understand its functionality)
@mjohnson541, could you comment?

davidfarinajr commented:
RMG-Py PR for adding G4 to Arkane: **adding G4 model chemistry to Arkane #1857**
`https://github.com/ReactionMechanismGenerator/RMG-Py/pull/1857`

rwest commented:
David - please could you write a bit more about the way the molecules were generated, and slightly more verbose details about how the thermo were calculated?
This is for me to paste into our progress report, but if you put it here for posterity it'll serve double purpose. Thanks.

davidfarinajr commented:
Sure,
1) I generated about ~1000 small (<=5 heavy atoms, so feasible for G4 method) CHON molecules (mult 1 and 2) using RMG.  I used the following reaction recipe to build closed shell molecules:
```
    recipe.actions = ([
    ['FORM_BOND', '*', o, '*'],
    ['LOSE_RADICAL', '*',str(o)],
    ])
```
where `o`  is a randomly chosen bond order.  I used isomorphism check as molecules were generated to ensure species are unique. To create radicals, I removed Hs from the closed shell molecules (that way I had complement saturated molecule to calculate HBIs).  The Jupyter notebook I used is here: `https://github.com/comocheng/halogen-data/blob/master/systematic_generation.ipynb` (sorry, it's a bit of a mess).
2) To calculate thermo, I used the systematic conformer generation scheme implemented in Auto-TST to find lowest energy conformer.  Conformers were generated by rotating dihedrals in 120 degree increments and optimized using `dftb+ (https://www.dftbplus.org/)` in `ASE` with the `halorg` parameter set (https://www.dftb.org/parameters/download/halorg/halorg-0-1-cc/) .  A diverse set of low energy conformers was kept with energy cutoff of `20 kcal/mol` and RMSD of `0.1 angstrom`.  The lowest energy conformer was identified by optimizing each of these low energy conformers at DFT level (`m062x-d3/jun-cc-pvtz`) in `Gaussian16`.  To make sure the molecule doesn't self-react or dissociate, I checked to see if it was isomorphic with starting molecule after the DFT or DFTb optimization.  Then, the energy of lowest energy conformer was evaluated with the Gaussian's `G4` method ( J. Chem. Phys. 126, 084108 (2007); doi: 10.1063/1.2436888). To determine the most representative Lewis Structure for molecules with resonance structures, I used natural bond orbital (NBO) analysis in Gaussian for closed-shell and the Huckel-Lewis Configuration Interaction method for open-shell (HL-CI) implemented by Alon in this notebook `https://github.com/alongd/HLCI`. Lastly, to get the thermo, I used Arkane with RRHO approx and bond additivity corrections derived from a set of ATcT training species.  
3) Benson groups were derived from the closed shell nitrogen molecules.  Groups that already had data were held fixed. Nitrogen groups that had no data or pointed to other groups were evaluated using least squares linear regression model. I also added some new groups by hand where needed.  I optimized about ~70 groups for 272 molecules.  Then, I generated ~200 new HBI values for the open shell species.

rwest commented:
Thanks!

1. How significant do you think is the error introduced by neglecting internal rotors (i.e. using RRHO instead of free or hindered rotors)?  And how expensive (your time, as well as CPU time) would it be to add? (I'm not sure how important this is, nor how fully automated it is).
2. Were species we generated in our HAN RMG runs included, or just these randomly generated ones? Perhaps your random sampling was thorough enough to get them all anyway (or all those ≤5 heavy atoms). 
3. The version of your notebook [here](https://github.com/comocheng/halogen-data/blob/master/systematic_generation.ipynb) stops at 4 heavy atoms, so never generates a quarternary molecule. Did you explore these?

alongd commented:
We have code for determining RRHO if you'd like. Here's a function to determine torsion pivots for a list of molecules of the same species (i.e., resonance structures):
https://github.com/ReactionMechanismGenerator/ARC/blob/master/arc/species/conformers.py#L668
We also have code to actually run them in ARC, and more importantly auto-troubeshoot failed scans. I'd happily elaborate over Slack.

davidfarinajr commented:
> * How significant do you think is the error introduced by neglecting internal rotors (i.e. using RRHO instead of free or hindered rotors)?  And how expensive (your time, as well as CPU time) would it be to add? (I'm not sure how important this is, nor how fully automated it is).

I'm not sure how significant this error is. Neglecting the internal rotors shouldn't affect **H298** much. However, it will have a more significant effect on **S** and **Cp**.  Therefore, neglecting this introduces more error in the free energy at higher temperatures. I think the error is on the order of several J/mol, whereas the error from a single point electronic energy calc, depending on the level of theory, can be several kcal/mol.  

Here is an example for ethane comparing entropy computed using RRHO vs Free vs Hindered rotor from `https://cccbdb.nist.gov/thermox.asp`  
Entropy is underestimated by about 1 J/mol K at 184 K using RRHO compared to experiment

<img width="1096" alt="Screen Shot 2020-01-03 at 1 30 48 PM" src="https://user-images.githubusercontent.com/32377555/71741637-6b7d4800-2e2d-11ea-8fc3-6a3afab67f02.png">

Adding rotor scans may shift the time bottleneck from the G4 single point calc to the rotor scans.  
G4 uses DFT (B3LYP) to optimize geometries and a series of single point calcs with CCSD(T) being the most expensive method. 

For example, For smiles `NNOO` running on 16 procs:
b3lyp opt: ~0.5 CPU hours
ccsd(t)/6-31G(d): ~7.5 CPU hours

If we do 3 rotor scans with 10 degree increments, the approx cost would be (assuming each geo opt takes about 0.5 CPU hours):
3 * 36 * 0.5 = 54 CPU hours

So it would be relatively expensive, but we could still do it. I am using `Tight` geo opt cutoffs in Gaussian, and we could save some time if we relax those.

I didn't do the rotor scans because I haven't automated it yet in my code.  If Alon's `ARC` can do it, I'll definitely look into that!

> * Were species we generated in our HAN RMG runs included, or just these randomly generated ones? Perhaps your random sampling was thorough enough to get them all anyway (or all those ≤5 heavy atoms).

Just the randomly generated ones.  I wanted to enumerate as many species as I could to derive estimation groups.  Hopefully we did catch some of RMG's species in the random sampling.  Moving forward, I'll focus on the species RMG generates.

> * The version of your notebook [here](https://github.com/comocheng/halogen-data/blob/master/systematic_generation.ipynb) stops at 4 heavy atoms, so never generates a quarternary molecule. Did you explore these?

Most of the species are <=4 heavy atoms.  However, If an N+ atom was randomly selected as the fourth heavy atom, it was paired with an O- atom to balance out the charge.  So my 5 heavy atom molecules all have N+/O- pairs, and therefore my sampling of 5 heavy atom space was less thorough than <5 heavy atom space.

davidfarinajr commented:
> We have code for determining RRHO if you'd like. Here's a function to determine torsion pivots for a list of molecules of the same species (i.e., resonance structures):
> https://github.com/ReactionMechanismGenerator/ARC/blob/master/arc/species/conformers.py#L668
> We also have code to actually run them in ARC, and more importantly auto-troubeshoot failed scans. I'd happily elaborate over Slack.

Thanks Alon!  I'll look into this and message you on Slack

davidfarinajr commented:
I need to do more work on derived the groups.  I think I overfit them to my training data.  I will open up another PR when I think its ready

alongd commented:
OK. You don't need to open a new PR, perhaps force push to this branch to maintain the discussion history?
